keps/sig-api-machinery/20190206-watch-bookmark.md

```diff
@@ -0,0 +1,180 @@
+from the last returned resourceVersion.
+
+This proposal make restarting watches cheaper from kube-apiserver performance
+perspective.
+
+## Motivation
+
+While running different scalability tests we observed that restarting watches
+may cause significant load on kube-apiserver when watcher is observing a small
+percentage of changes (due to field or label selector). In extreme cases,
+reestablishing such watcher may even lead to falling out of history window
+and "resource version too old" errors (that requires full relist for that
+watcher).
+
+The reason for that is the fact that even if the last item received by watcher
+has resourceVersion rv1, we may already know that there aren't any changes
+a given watcher is interested in up to rv2 (rv1 < rv2), but we don't have any
+way of communicating it to the watcher. As a result, when restarting a watch,
+client again sends rv1 as a starting point, and we process all events with
+resourceVersion between rv1 and rv2 again unnecessarily.
+
+The proposal presents a proper solution for that problem.
+
+### Goals
+
+- Reduce load on apiserver by minimizing amount of unnecessary watch events
+that need to be processed after restarting a watch.
+- Reduce amount of undesired "resource version too old" errors on reestablishing
+a watch.
+
+### Non-Goals
+
+The following are nice-to-haves, but not primary goals:
+
+- Improve overall watch throughput and/or latency.
+
+## Proposal
+
+We propose introducing a new type of watch event called `Bookmark`. With that
+change, the possible watch event types will be:
+```
+  Added    EventType = "ADDED"
+  Modified EventType = "MODIFIED"
+  Deleted  EventType = "DELETED"
+  Error    EventType = "ERROR"
+  Bookmark EventType = "BOOKMARK"
+```
+
+Watch event with type Bookmark will represent information that all the objects
+up to a given resourceVersion has been processed for a given watcher. So even
+if the last event of other types contained object with resourceVersion rv1,
+receiving a bookmark with resourceVersion rv2 means that there aren't
+any interesting objects for that watcher in between.
+
+Given that we don't want to wait for v2 version of watch API with that change,
+an obvious requirement for introducing it is backward compatibility. Currently
+watch Event type looks as following:
+```
+  type Event struct {
+    Type EventType
+    Object runtime.Object
+  }
+```
+
+As a result, we will represent bookmark event by setting Bookmark type and
+Object of appropriate type with just ObjectMeta.ResourceVersion field set.
+
+Unfortunately, such a change would break existing clients, as an example
+[official decoder][]. As a result, we will extend `ListOptions` (which is
+how we pass options to watch) with a boolean field where user can opt-in
+for watch bookmarks:
+```
+  type ListOptions struct {
+    ...
+    RequestBookmarks bool
+  }
+```
+We consciously make it just a boolean flag - this gives kube-apiserver an
```

Comment:
Also explain that this makes backwards compatibility safe because the server will drop it during parsing (you'll need to confirm that, but I'm 95% sure) and the client isn't guaranteed to get any bookmarks

Reply:
You mean that the old apiserver will drop this new RequestBookmarks field?
If so, then yes, it will be dropped, I already checked that and added to the test.
If you had something else on your mind, can you please clarify?

Reply:
> You mean that the old apiserver will drop this new RequestBookmarks field?
> If so, then yes, it will be dropped, I already checked that and added to the test.
> If you had something else on your mind, can you please clarify?

Yes, that's what I'm thinking.  I suspect others may have the question, but won't know how the machinery works